[Main] Publish Artifacts step failed after rerun

Related to https://github.com/dotnet/source-build/issues/4537 and https://github.com/dotnet/sdk/commit/e5a8c453ae2cc837609126562e12e7e914e39752, it seems that @ViktorHofer added the condition `succeededOrFailed()` to the publish artifacts step and it does not support overwriting artifacts.

[Failing build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2620900&view=logs&j=ddffac02-ce19-58a3-8fe4-a1a93b2827d8&t=e392a9fd-ea1a-55e4-f224-f244f69eff09) (internal Microsoft link):

    ##[error]Artifact CentOSStream9_Mono_Offline_CurrentSourceBuiltSdk_x64_Artifacts already exists for build 2620900.

Does below solution make sense to you @MichaelSimons @ellahathaway @ViktorHofer  ?

- Add another Publish Artifacts step that only publish on success, and modify the current Publish Artifacts step to include `$(System.JobAttempt)` that only publish on fail. This will not affect the other pipelines download these artifacts.

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -627,7 +627,7 @@ jobs:
 
     - publish: $(artifactsStagingDir)
       artifact: $(failedJobArtifactName)
-      displayName: Publish Artifacts
+      displayName: Publish Artifacts (On Failure)
       condition: failed()
       continueOnError: true
 

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -621,7 +621,7 @@ jobs:
     # This prevents overwrite conflicts in the event of a failed build followed by a rerun and a successful build.
     - publish: $(artifactsStagingDir)
       artifact: $(successfulJobArtifactName)
-      displayName: Publish Artifacts
+      displayName: Publish Artifacts (On Success)
       condition: succeeded()
       continueOnError: true
 

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -214,7 +214,7 @@ jobs:
     - output: pipelineArtifact
       path: $(artifactsStagingDir)
       artifact: $(successfulJobArtifactName)
-      displayName: Publish Artifacts
+      displayName: Publish Artifacts (On Success)
       condition: succeeded()
       sbomEnabled: true
 

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -209,6 +209,8 @@ jobs:
       artifactName: $(Agent.JobName)_BuildLogs_Attempt$(System.JobAttempt)
       sbomEnabled: false
 
+    # Both publishing steps are necessary to ensure artifacts are published on both success and failure.
+    # This prevents overwrite conflicts in the event of a failed build followed by a rerun and a successful build.
     - output: pipelineArtifact
       path: $(artifactsStagingDir)
       artifact: $(successfulJobArtifactName)
@@ -615,6 +617,8 @@ jobs:
         testRunTitle: ScenarioTests_$(Agent.JobName)
 
   - ${{ if or(ne(variables['System.TeamProject'], 'internal'), eq(variables['Build.Reason'], 'PullRequest')) }}:
+    # Both publishing steps are necessary to ensure artifacts are published on both success and failure.
+    # This prevents overwrite conflicts in the event of a failed build followed by a rerun and a successful build.
     - publish: $(artifactsStagingDir)
       artifact: $(successfulJobArtifactName)
       displayName: Publish Artifacts

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -210,7 +210,8 @@ jobs:
       sbomEnabled: false
 
     # Both publishing steps are necessary to ensure artifacts are published on both success and failure.
-    # This prevents overwrite conflicts in the event of a failed build followed by a rerun and a successful build.
+    # This prevents overwrite conflicts in the event of a failed build followed by a rerun.
+    # Additionally, the 'Download Previous Build *' steps depend on a fixed name to acquire specific assets in multi-stage builds.
     - output: pipelineArtifact
       path: $(artifactsStagingDir)
       artifact: $(successfulJobArtifactName)
@@ -619,6 +620,7 @@ jobs:
   - ${{ if or(ne(variables['System.TeamProject'], 'internal'), eq(variables['Build.Reason'], 'PullRequest')) }}:
     # Both publishing steps are necessary to ensure artifacts are published on both success and failure.
     # This prevents overwrite conflicts in the event of a failed build followed by a rerun.
+    # Additionally, the 'Download Previous Build *' steps depend on a fixed name to acquire specific assets in multi-stage builds.
     - publish: $(artifactsStagingDir)
       artifact: $(successfulJobArtifactName)
       displayName: Publish Artifacts (On Success)

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -221,7 +221,7 @@ jobs:
     - output: pipelineArtifact
       path: $(artifactsStagingDir)
       artifact: $(failedJobArtifactName)
-      displayName: Publish Artifacts
+      displayName: Publish Artifacts (On Failure)
       condition: failed()
       sbomEnabled: true
 

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -191,7 +191,7 @@ jobs:
     value: $(Agent.JobName)_Artifacts
 
   - name: failedJobArtifactName
-    value: $(Agent.JobName)_Artifacts_Attempt$(System.JobAttempt)
+    value: $(successfulJobArtifactName)_Attempt$(System.JobAttempt)
 
   # manually disable CodeQL until https://dev.azure.com/mseng/1ES/_workitems/edit/2211548 is implemented
   # CodeQL doesn't work on arm64 macOS, see https://portal.microsofticm.com/imp/v5/incidents/details/532165079/summary

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -618,7 +618,7 @@ jobs:
 
   - ${{ if or(ne(variables['System.TeamProject'], 'internal'), eq(variables['Build.Reason'], 'PullRequest')) }}:
     # Both publishing steps are necessary to ensure artifacts are published on both success and failure.
-    # This prevents overwrite conflicts in the event of a failed build followed by a rerun and a successful build.
+    # This prevents overwrite conflicts in the event of a failed build followed by a rerun.
     - publish: $(artifactsStagingDir)
       artifact: $(successfulJobArtifactName)
       displayName: Publish Artifacts (On Success)

--- a/eng/pipelines/templates/jobs/vmr-build.yml
+++ b/eng/pipelines/templates/jobs/vmr-build.yml
@@ -187,6 +187,12 @@ jobs:
   - name: artifactsStagingDir
     value: $(Build.ArtifactStagingDirectory)/artifacts
 
+  - name: successfulJobArtifactName
+    value: $(Agent.JobName)_Artifacts
+
+  - name: failedJobArtifactName
+    value: $(Agent.JobName)_Artifacts_Attempt$(System.JobAttempt)
+
   # manually disable CodeQL until https://dev.azure.com/mseng/1ES/_workitems/edit/2211548 is implemented
   # CodeQL doesn't work on arm64 macOS, see https://portal.microsofticm.com/imp/v5/incidents/details/532165079/summary
   - ${{ if eq(parameters.pool.os, 'macOS') }}:
@@ -205,9 +211,16 @@ jobs:
 
     - output: pipelineArtifact
       path: $(artifactsStagingDir)
-      artifact: $(Agent.JobName)_Artifacts
+      artifact: $(successfulJobArtifactName)
       displayName: Publish Artifacts
-      condition: succeededOrFailed()
+      condition: succeeded()
+      sbomEnabled: true
+
+    - output: pipelineArtifact
+      path: $(artifactsStagingDir)
+      artifact: $(failedJobArtifactName)
+      displayName: Publish Artifacts
+      condition: failed()
       sbomEnabled: true
 
     # Using build artifacts to enable publishing the vertical manifests to a single artifact from different jobs
@@ -603,9 +616,15 @@ jobs:
 
   - ${{ if or(ne(variables['System.TeamProject'], 'internal'), eq(variables['Build.Reason'], 'PullRequest')) }}:
     - publish: $(artifactsStagingDir)
-      artifact: $(Agent.JobName)_Artifacts
+      artifact: $(successfulJobArtifactName)
       displayName: Publish Artifacts
-      condition: succeededOrFailed()
+      condition: succeeded()
+      continueOnError: true
+
+    - publish: $(artifactsStagingDir)
+      artifact: $(failedJobArtifactName)
+      displayName: Publish Artifacts
+      condition: failed()
       continueOnError: true
 
     # Using build artifacts to enable publishing the vertical manifests to a single artifact from different jobs

